File: DSA/LeetCode/DFS/1382. Balance a Binary Search Tree.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var balanceBST = function (root) {


    function dfs(node) {
        if (!node)
            return []

        return [...dfs(node.left), node.val, ...dfs(node.right)];
    }
    const res = dfs(root);

    function rec(start, end) {
        if (start > end) {
            return null;
        }

        let mid = Math.floor((start + end) / 2);


        return new TreeNode(
            res[mid],
            rec(start, mid - 1),
            rec(mid + 1, end)
        );
    }

    return rec(0, res.length - 1)

};

File: DSA/LeetCode/DFS/1382. Balance a Binary Search Tree.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def balanceBST(self, root: TreeNode) -> TreeNode:

       
        def dfs(node):
            if not node:
                return []
            return dfs(node.left) + [node.val] + dfs(node.right)

        res = dfs(root)

        def rec(start, end):
            if start > end:
                return None
            mid = (start + end) // 2
            return TreeNode(
                res[mid],
                rec(start, mid - 1),
                rec(mid + 1, end)
            )

        return rec(0, len(res) - 1)
        